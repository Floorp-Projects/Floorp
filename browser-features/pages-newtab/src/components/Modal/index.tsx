import { useEffect, useRef } from "react";
import { useTranslation } from "react-i18next";

interface ModalProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
  title: string;
}

export function Modal({ isOpen, onClose, children, title }: ModalProps) {
  const { t } = useTranslation();
  const overlayRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape" && isOpen) {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    return () => document.removeEventListener("keydown", handleEscape);
  }, [isOpen, onClose]);

  if (!isOpen) return null;

  const handleOverlayClick = (e: React.MouseEvent) => {
    if (e.target === overlayRef.current) {
      onClose();
    }
  };

  return (
    <div
      className="modal modal-open"
      ref={overlayRef}
      onClick={handleOverlayClick}
    >
      <div className="modal-box max-w-2xl">
        <div className="flex justify-between items-center">
          <h3 className="font-bold text-lg">{title}</h3>
          <button
            type="button"
            onClick={onClose}
            className="btn btn-sm btn-circle btn-ghost"
          >
            <span className="sr-only">{t("modal.close")}</span>
            <svg
              className="h-5 w-5"
              fill="none"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <path d="M6 18L18 6M6 6l12 12" />
            </svg>
          </button>
        </div>
        <div className="py-4">
          {children}
        </div>
        <div className="modal-action">
          <button onClick={onClose} className="btn bg-primary text-primary-content">{t("modal.close")}</button>
        </div>
      </div>
    </div>
  );
}
