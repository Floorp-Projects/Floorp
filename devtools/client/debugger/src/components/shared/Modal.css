/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

.modal-wrapper {
  position: fixed;
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  transition: z-index 200ms;
  z-index: 100;
}

.modal {
  display: flex;
  width: 80%;
  max-height: 80vh;
  overflow-y: auto;
  background-color: var(--theme-toolbar-background);
  transition: transform 150ms cubic-bezier(0.07, 0.95, 0, 1);
  box-shadow: 1px 1px 6px 1px var(--popup-shadow-color);
}

.modal.entering,
.modal.exited {
  transform: translateY(-101%);
}

.modal.entered,
.modal.exiting {
  transform: translateY(5px);
  flex-direction: column;
}

/* This rule is active when the screen is not narrow */
@media (min-width: 580px) {
  .modal {
    width: 50%;
  }
}

@media (min-height: 340px) {
  .modal.entered,
  .modal.exiting {
    transform: translateY(30px);
  }
}
