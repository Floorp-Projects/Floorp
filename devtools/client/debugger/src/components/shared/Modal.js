/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import PropTypes from "prop-types";
import React from "react";
import classnames from "classnames";
import Transition from "react-transition-group/Transition";
import "./Modal.css";

export const transitionTimeout = 50;

export class Modal extends React.Component {
  onClick = e => {
    e.stopPropagation();
  };

  render() {
    const { additionalClass, children, handleClose, status } = this.props;

    return (
      <div className="modal-wrapper" onClick={handleClose}>
        <div
          className={classnames("modal", additionalClass, status)}
          onClick={this.onClick}
        >
          {children}
        </div>
      </div>
    );
  }
}

Modal.contextTypes = {
  shortcuts: PropTypes.object,
};

export default function Slide({
  in: inProp,
  children,
  additionalClass,
  handleClose,
}) {
  return (
    <Transition in={inProp} timeout={transitionTimeout} appear>
      {status => (
        <Modal
          status={status}
          additionalClass={additionalClass}
          handleClose={handleClose}
        >
          {children}
        </Modal>
      )}
    </Transition>
  );
}
