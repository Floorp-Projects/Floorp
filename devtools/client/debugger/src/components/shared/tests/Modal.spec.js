/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { shallow } from "enzyme";

import { Modal } from "../Modal";

describe("Modal", () => {
  it("renders", () => {
    const wrapper = shallow(<Modal handleClose={() => {}} status="entering" />);
    expect(wrapper).toMatchSnapshot();
  });

  it("handles close modal click", () => {
    const handleCloseSpy = jest.fn();
    const wrapper = shallow(
      <Modal handleClose={handleCloseSpy} status="entering" />
    );
    wrapper.find(".modal-wrapper").simulate("click");
    expect(handleCloseSpy).toHaveBeenCalled();
  });

  it("renders children", () => {
    const children = <div className="aChild" />;
    const wrapper = shallow(
      <Modal children={children} handleClose={() => {}} status="entering" />
    );
    expect(wrapper.find(".aChild")).toHaveLength(1);
  });

  it("passes additionalClass to child div class", () => {
    const additionalClass = "testAddon";
    const wrapper = shallow(
      <Modal
        additionalClass={additionalClass}
        handleClose={() => {}}
        status="entering"
      />
    );
    expect(wrapper.find(`.modal-wrapper .${additionalClass}`)).toHaveLength(1);
  });

  it("passes status to child div class", () => {
    const status = "testStatus";
    const wrapper = shallow(<Modal status={status} handleClose={() => {}} />);
    expect(wrapper.find(`.modal-wrapper .${status}`)).toHaveLength(1);
  });
});
