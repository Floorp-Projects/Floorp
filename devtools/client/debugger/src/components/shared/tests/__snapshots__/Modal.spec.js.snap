// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders 1`] = `
<div
  className="modal-wrapper"
  onClick={[Function]}
>
  <div
    className="modal entering"
    onClick={[Function]}
  />
</div>
`;
